Support older twinkly devices without effects

Olen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This should bring back support for older devices that was broken with the new effects functions.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #83076
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Olen committed:
Support older devices without effects

Olen committed:
Also try around turn_on-functions

Olen committed:
Fix typos

Olen committed:
Remove print-statement

Olen committed:
Improving error-logging

Olen committed:
Trying even harder to detect old versions

Olen committed:
Using version string to detect effect support

firstof9 commented:
Confirmed with local testing on Twinkly device firmware version: `2.3.5`

Olen commented:
Confirmed to fix https://github.com/home-assistant/core/issues/83076

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
-                            self._attr_rgbw_color[0],
-                            self._attr_rgbw_color[1],
-                            self._attr_rgbw_color[2],
+                    if LightEntityFeature.EFFECT in self.supported_features:
```
`in` is not the correct check. `supported_features` is an integer. Calling `in` on an integer raises a `TypeError`.

balloob replied:
Ah, I see now that this is supported because supported features inherits from `IntFlag`.

Olen replied:
Sorry, thought it was a LightEntityFeature, which again is a enum.IntFlag.  
Changed to & now throughout the code for consitency.

Olen replied:
Missed your second comment.  Should I add it back as "in" or do you prefer &?
Personally I think "in" is easier to read, but that is just my opinion.

balloob replied:
`in` is easier to read once you understand it's a flag 😅 I don't really care.

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+                        await self._client.set_mode("movie")
+                        self._client.default_mode = "movie"
 
         if ATTR_RGB_COLOR in kwargs:
             if kwargs[ATTR_RGB_COLOR] != self._attr_rgb_color:
```
These if-statements should be merged to save some indentation.

Olen replied:
Fixed

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+                                await self._client.set_mode("movie")
+                                self._client.default_mode = "movie"
+                            else:
+                                _LOGGER.warning("Unable to set rgbw-color")
+                                _LOGGER.warning(error)
```
Single log statement. 

```suggestion
                                _LOGGER.warning("Unable to set rgbw-color: %s", error)
```

Olen replied:
Removed the logging as we now rely on software-version instead of http-errors to detect supported features

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+                            await self._client.set_mode("color")
+                            self._client.default_mode = "color"
+                        except ClientResponseError as error:
+                            if error.status == HTTPNotFound().status:
```
Use the constant for 404, don't misuse an error exception.

Olen replied:
Removed, as we now rely on software-version instead of http-errors to detect supported features

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
                 if isinstance(self._attr_rgbw_color, tuple):
```
What could it be that's not a tuple ? You just assigned it and it came from kwargs. We would never pass a non-tuple.

balloob replied:
Also, you should only assign it to `self._attr_rgbw_color` after the call was a success. Otherwise your locally cached state of the device is wrong.

Olen replied:
Fixed

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -185,16 +185,47 @@ async def async_turn_on(self, **kwargs: Any) -> None:
+                                self._attr_supported_features = (
+                                    self.supported_features & ~LightEntityFeature.EFFECT
+                                )
+                                await self._client.set_cycle_colours(
```
This is an exact copy of the code of when the feature is not supported. That's not good. Better to just call yourself again.

```python
await self.async_turn_on(call)
return
```

Olen replied:
Removed as we now rely on software-version instead of http-errors to detect supported features

balloob commented on tests/components/twinkly/test_light.py:
```diff
@@ -171,6 +171,110 @@ async def test_turn_on_with_effect(hass: HomeAssistant):
+    assert hass.states.get(entity.entity_id).state == "off"
+    assert not client.current_movie
+
+    assert LightEntityFeature.EFFECT in entity.supported_features
```
How is this test passing? `entity.supportd_features` is an int field. 

```
>>> LightEntityFeature.EFFECT in 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: argument of type 'int' is not iterable
```

balloob replied:
nm. looks like supported feature inherits from `Intflag` now.

balloob commented on tests/components/twinkly/test_light.py:
```diff
@@ -171,6 +171,110 @@ async def test_turn_on_with_effect(hass: HomeAssistant):
+    client.mock_unsupported_effects = True
+
+    # Turn off to force async_update
+    await hass.services.async_call(
+        "light", "turn_off", service_data={"entity_id": entity.entity_id}
+    )
+    await hass.async_block_till_done()
```
```suggestion
    await hass.services.async_call(
        "light", "turn_off", service_data={"entity_id": entity.entity_id}, blocking=True
    )
```

Olen replied:
Thank you.  Updated all the tests to use this.

balloob commented on tests/components/twinkly/test_light.py:
```diff
@@ -171,6 +171,110 @@ async def test_turn_on_with_effect(hass: HomeAssistant):
+        "turn_on",
+        service_data={"entity_id": entity.entity_id, "rgbw_color": (128, 64, 32, 0)},
+    )
+    await hass.async_block_till_done()
```
same

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -268,8 +321,20 @@ async def async_update(self) -> None:
 
-            await self.async_update_movies()
-            await self.async_update_current_movie()
+            if ATTR_VERSION not in self._attributes:
```
We shouldn't add this to the state attributes. It should be be returned from  device_info

Olen replied:
Fixed

balloob commented on homeassistant/components/twinkly/light.py:
```diff
@@ -96,6 +100,7 @@ def __init__(
         self._attributes: dict[Any, Any] = {}
         self._current_movie: dict[Any, Any] = {}
         self._movies: list[Any] = []
+        self._attr_supported_features = LightEntityFeature.EFFECT
```
You should set this by checking the firmware version against `MIN_EFFECT_VERSION` in `async_added_to_hass` life cycle.

Olen replied:
Fixed

Olen committed:
Code cleanup and fixes from review

Olen committed:
Remove extra version check

Olen committed:
Cleanup and improve tests

Olen commented:
Thank you for the review.
All the issues have been fixed, I believe.  I removed a lot of code and now only use the detected software-version to disable effects.
Also cleaned up a lot of the tests.

@firstof9 If you have time, could you please validate that this still works on FW < 2.7.1?

Olen committed:
Remove unused "max_movies"

Olen committed:
Remove extra tests when not setting effect

Olen committed:
Also test setting effect when effect is not supported

Olen committed:
Fix docstring

firstof9 commented:
Confirmed working on my Twinkly device.

ludeeus commented on homeassistant/components/twinkly/light.py:
```diff
@@ -7,6 +7,7 @@
 from typing import Any
 
 from aiohttp import ClientError
+from packaging import version
```
Why is this not using awesomeversion like all other integrations/parts?

Olen replied:
Just because I did not know all other integrations use that.  
I'll add another PR to switch

Olen replied:
https://github.com/home-assistant/core/pull/83205

ludeeus replied:
Great 👍
